<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>Identify Copy Number Variations (CNVs) in cancer cells</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Identify Copy Number Variations (CNVs) in
cancer cells</h1>



<p>Analyze snATAC-seq data of basal cell carcinoma sample
SU008_Tumor_Pre in GEO (GSE129785).</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(scPloidy)</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="co">#&gt; Attaching package: &#39;dplyr&#39;</span></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="co">#&gt; The following objects are masked from &#39;package:stats&#39;:</span></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a><span class="co">#&gt;     filter, lag</span></span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a><span class="co">#&gt; The following objects are masked from &#39;package:base&#39;:</span></span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a><span class="fu">library</span>(tidyr)</span>
<span id="cb1-12"><a href="#cb1-12" tabindex="-1"></a><span class="fu">library</span>(gplots)</span>
<span id="cb1-13"><a href="#cb1-13" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb1-14"><a href="#cb1-14" tabindex="-1"></a><span class="co">#&gt; Attaching package: &#39;gplots&#39;</span></span>
<span id="cb1-15"><a href="#cb1-15" tabindex="-1"></a><span class="co">#&gt; The following object is masked from &#39;package:stats&#39;:</span></span>
<span id="cb1-16"><a href="#cb1-16" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb1-17"><a href="#cb1-17" tabindex="-1"></a><span class="co">#&gt;     lowess</span></span></code></pre></div>
<p>You can skip the preprocessing and start from section CNV.</p>
<div id="download-from-geo" class="section level2">
<h2>Download from GEO</h2>
<p>Download GSE129785_scATAC-TME-All.cell_barcodes.txt.gz from below and
gunzip <a href="https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE129785&amp;format=file&amp;file=GSE129785%5FscATAC%2DTME%2DAll%2Ecell%5Fbarcodes%2Etxt%2Egz" class="uri">https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE129785&amp;format=file&amp;file=GSE129785%5FscATAC%2DTME%2DAll%2Ecell%5Fbarcodes%2Etxt%2Egz</a></p>
<p>Download GSM3722064_SU008_Tumor_Pre_fragments.tsv.gz from <a href="https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSM3722064&amp;format=file&amp;file=GSM3722064%5FSU008%5FTumor%5FPre%5Ffragments%2Etsv%2Egz" class="uri">https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSM3722064&amp;format=file&amp;file=GSM3722064%5FSU008%5FTumor%5FPre%5Ffragments%2Etsv%2Egz</a></p>
</div>
<div id="prepare-windows-and-peaks" class="section level2">
<h2>Prepare windows and peaks</h2>
<p>The input window file <code>window.hg37.20MB.bed</code> and resultant
peak file <code>multi_tissue_peaks.hg37.20MB.bed</code> can be
downloaded from <a href="https://doi.org/10.6084/m9.figshare.23574066" class="uri">https://doi.org/10.6084/m9.figshare.23574066</a></p>
<p>To reproduce by yourself, download chromatin accessibility
DHS_Index_and_Vocabulary_hg19_WM20190703.txt.gz from <a href="https://doi.org/10.5281/zenodo.3838751" class="uri">https://doi.org/10.5281/zenodo.3838751</a></p>
<p>Generate peaks for 20MB windows using peak_sum.R by yardimcilab in <a href="https://github.com/yardimcilab/RIDDLER/blob/main/util/peak_sum.R" class="uri">https://github.com/yardimcilab/RIDDLER/blob/main/util/peak_sum.R</a></p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a>SU008_Tumor_Pre_windowcovariates <span class="ot">=</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>  <span class="fu">read.table</span>(</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a>    <span class="st">&quot;multi_tissue_peaks.hg37.20MB.bed&quot;</span>,</span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a>    <span class="at">header =</span> <span class="cn">FALSE</span>)</span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a><span class="fu">colnames</span>(SU008_Tumor_Pre_windowcovariates) <span class="ot">=</span></span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a>  <span class="fu">c</span>(<span class="st">&quot;chr&quot;</span>, <span class="st">&quot;start&quot;</span>, <span class="st">&quot;end&quot;</span>, <span class="st">&quot;window&quot;</span>, <span class="st">&quot;peaks&quot;</span>)</span></code></pre></div>
</div>
<div id="compute-fragmentoverlap-from-fragments" class="section level2">
<h2>Compute fragmentoverlap from fragments</h2>
<p>See vignette of R package scPloidy. Load setting for hg19 genome.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a>simpleRepeat <span class="ot">=</span> readr<span class="sc">::</span><span class="fu">read_tsv</span>(</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>  <span class="st">&quot;~/human/publichuman/hg37_ucsc/simpleRepeat.chrom_chromStart_chromEnd.txt.gz&quot;</span>,</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>  <span class="at">col_names =</span> <span class="fu">c</span>(<span class="st">&quot;chrom&quot;</span>, <span class="st">&quot;chromStart&quot;</span>, <span class="st">&quot;chromEnd&quot;</span>))</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>rmsk <span class="ot">=</span> readr<span class="sc">::</span><span class="fu">read_tsv</span>(</span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a>  <span class="st">&quot;~/human/publichuman/hg37_ucsc/rmsk.Simple_repeat.genoName_genoStart_genoEnd.txt.gz&quot;</span>,</span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a>  <span class="at">col_names =</span> <span class="fu">c</span>(<span class="st">&quot;chrom&quot;</span>, <span class="st">&quot;chromStart&quot;</span>, <span class="st">&quot;chromEnd&quot;</span>))</span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a>simpleRepeat <span class="ot">=</span> <span class="fu">rbind</span>(simpleRepeat, rmsk)</span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a><span class="fu">rm</span>(rmsk)</span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a><span class="co"># convert from 0-based position to 1-based</span></span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a>simpleRepeat[, <span class="dv">2</span>] <span class="ot">=</span> simpleRepeat[, <span class="dv">2</span>] <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a>simpleRepeat <span class="ot">=</span> GenomicRanges<span class="sc">::</span><span class="fu">makeGRangesFromDataFrame</span>(</span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a>  <span class="fu">as.data.frame</span>(simpleRepeat),</span>
<span id="cb3-13"><a href="#cb3-13" tabindex="-1"></a>  <span class="at">seqnames.field =</span> <span class="st">&quot;chrom&quot;</span>,</span>
<span id="cb3-14"><a href="#cb3-14" tabindex="-1"></a>  <span class="at">start.field    =</span> <span class="st">&quot;chromStart&quot;</span>,</span>
<span id="cb3-15"><a href="#cb3-15" tabindex="-1"></a>  <span class="at">end.field      =</span> <span class="st">&quot;chromEnd&quot;</span>)</span>
<span id="cb3-16"><a href="#cb3-16" tabindex="-1"></a><span class="co"># remove duplicates</span></span>
<span id="cb3-17"><a href="#cb3-17" tabindex="-1"></a>simpleRepeat <span class="ot">=</span> GenomicRanges<span class="sc">::</span><span class="fu">union</span>(simpleRepeat, GenomicRanges<span class="sc">::</span><span class="fu">GRanges</span>())</span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a>window <span class="ot">=</span> <span class="fu">read.table</span>(<span class="st">&quot;window.hg37.20MB.bed&quot;</span>, <span class="at">header =</span> <span class="cn">FALSE</span>)</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="fu">colnames</span>(window) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&quot;chr&quot;</span>, <span class="st">&quot;start&quot;</span>, <span class="st">&quot;end&quot;</span>, <span class="st">&quot;window&quot;</span>)</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a>at <span class="ot">=</span> GenomicRanges<span class="sc">::</span><span class="fu">makeGRangesFromDataFrame</span>(window[, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>])</span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a>barcodesuffix <span class="ot">=</span> <span class="fu">paste0</span>(<span class="st">&quot;.&quot;</span>, window<span class="sc">$</span>window)</span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a>sc <span class="ot">=</span> <span class="fu">read.csv</span>(</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>  <span class="st">&quot;GSE129785_scATAC-TME-All.cell_barcodes.txt&quot;</span>,</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a>  <span class="at">header =</span> <span class="cn">TRUE</span>,</span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a>  <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>)</span></code></pre></div>
<p>Compute and save fragmentoverlap.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a>sample <span class="ot">=</span> <span class="st">&quot;GSM3722064&quot;</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>tissue <span class="ot">=</span> <span class="st">&quot;SU008_Tumor_Pre&quot;</span></span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a>bc <span class="ot">=</span> sc<span class="sc">$</span>Barcodes[sc<span class="sc">$</span>Group <span class="sc">==</span> tissue]</span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a>SU008_Tumor_Pre_fragmentoverlap <span class="ot">=</span></span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a>  <span class="fu">fragmentoverlapcount</span>(</span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a>    <span class="fu">paste0</span>(<span class="st">&quot;SRX5679934/&quot;</span>, sample, <span class="st">&quot;_&quot;</span>, tissue, <span class="st">&quot;_fragments.tsv.gz&quot;</span>),</span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a>    at,</span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a>    <span class="at">excluderegions =</span> simpleRepeat,</span>
<span id="cb6-10"><a href="#cb6-10" tabindex="-1"></a>    <span class="at">targetbarcodes =</span> bc,</span>
<span id="cb6-11"><a href="#cb6-11" tabindex="-1"></a>    <span class="at">Tn5offset =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">0</span>),</span>
<span id="cb6-12"><a href="#cb6-12" tabindex="-1"></a>    <span class="at">barcodesuffix =</span> barcodesuffix</span>
<span id="cb6-13"><a href="#cb6-13" tabindex="-1"></a>  )</span></code></pre></div>
</div>
<div id="cnv" class="section level2">
<h2>CNV</h2>
<p>You can skip above and load preprocessed data attached to the
package. The data file GSE129785_SU008_Tumor_Pre.RData is also available
from <a href="https://doi.org/10.6084/m9.figshare.23574066" class="uri">https://doi.org/10.6084/m9.figshare.23574066</a></p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="fu">data</span>(GSE129785_SU008_Tumor_Pre)</span></code></pre></div>
<p>Infer CNVs.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a>levels <span class="ot">=</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">4</span>)</span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>result <span class="ot">=</span> <span class="fu">cnv</span>(SU008_Tumor_Pre_fragmentoverlap,</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a>             SU008_Tumor_Pre_windowcovariates,</span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a>             <span class="at">levels =</span> levels,</span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a>             <span class="at">deltaBICthreshold =</span> <span class="sc">-</span><span class="dv">600</span>)</span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a><span class="co">#&gt; [1] &quot;Computing span = 2&quot;</span></span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a><span class="co">#&gt; [1] &quot;Computing span = 3&quot;</span></span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a><span class="co">#&gt; [1] &quot;Computing span = 4&quot;</span></span>
<span id="cb8-9"><a href="#cb8-9" tabindex="-1"></a><span class="co">#&gt; [1] &quot;Computing span = 5&quot;</span></span>
<span id="cb8-10"><a href="#cb8-10" tabindex="-1"></a><span class="co">#&gt; [1] &quot;Computing span = 6&quot;</span></span>
<span id="cb8-11"><a href="#cb8-11" tabindex="-1"></a><span class="co">#&gt; [1] &quot;Computing span = 7&quot;</span></span>
<span id="cb8-12"><a href="#cb8-12" tabindex="-1"></a><span class="co">#&gt; [1] &quot;Computing span = 8&quot;</span></span>
<span id="cb8-13"><a href="#cb8-13" tabindex="-1"></a><span class="co">#&gt; [1] &quot;Computing span = 9&quot;</span></span>
<span id="cb8-14"><a href="#cb8-14" tabindex="-1"></a><span class="co">#&gt; [1] &quot;Computing span = 10&quot;</span></span>
<span id="cb8-15"><a href="#cb8-15" tabindex="-1"></a><span class="co">#&gt; [1] &quot;Computing span = 2&quot;</span></span>
<span id="cb8-16"><a href="#cb8-16" tabindex="-1"></a><span class="co">#&gt; [1] &quot;Computing span = 3&quot;</span></span>
<span id="cb8-17"><a href="#cb8-17" tabindex="-1"></a><span class="co">#&gt; [1] &quot;Computing span = 4&quot;</span></span>
<span id="cb8-18"><a href="#cb8-18" tabindex="-1"></a><span class="co">#&gt; [1] &quot;Computing span = 5&quot;</span></span>
<span id="cb8-19"><a href="#cb8-19" tabindex="-1"></a><span class="co">#&gt; [1] &quot;Computing span = 6&quot;</span></span>
<span id="cb8-20"><a href="#cb8-20" tabindex="-1"></a><span class="co">#&gt; [1] &quot;Computing span = 7&quot;</span></span>
<span id="cb8-21"><a href="#cb8-21" tabindex="-1"></a><span class="co">#&gt; [1] &quot;Computing span = 8&quot;</span></span>
<span id="cb8-22"><a href="#cb8-22" tabindex="-1"></a><span class="co">#&gt; [1] &quot;Computing span = 9&quot;</span></span>
<span id="cb8-23"><a href="#cb8-23" tabindex="-1"></a><span class="co">#&gt; [1] &quot;Computing span = 10&quot;</span></span></code></pre></div>
<p>Attach the result to <code>fragmentoverlap</code>.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a>windowcovariates <span class="ot">=</span> SU008_Tumor_Pre_windowcovariates</span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a>windowcovariates<span class="sc">$</span>w <span class="ot">=</span></span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a>  <span class="fu">as.numeric</span>(<span class="fu">sub</span>(<span class="st">&quot;window_&quot;</span>, <span class="st">&quot;&quot;</span>, windowcovariates<span class="sc">$</span>window))</span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a></span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a>fragmentoverlap <span class="ot">=</span> SU008_Tumor_Pre_fragmentoverlap</span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a>fragmentoverlap<span class="sc">$</span>cell <span class="ot">=</span></span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a>  <span class="fu">sub</span>(<span class="st">&quot;.window.*&quot;</span>, <span class="st">&quot;&quot;</span>, fragmentoverlap<span class="sc">$</span>barcode)</span>
<span id="cb9-8"><a href="#cb9-8" tabindex="-1"></a>fragmentoverlap<span class="sc">$</span>window <span class="ot">=</span></span>
<span id="cb9-9"><a href="#cb9-9" tabindex="-1"></a>  <span class="fu">sub</span>(<span class="st">&quot;.*window&quot;</span>, <span class="st">&quot;window&quot;</span>, fragmentoverlap<span class="sc">$</span>barcode)</span>
<span id="cb9-10"><a href="#cb9-10" tabindex="-1"></a>fragmentoverlap<span class="sc">$</span>w <span class="ot">=</span></span>
<span id="cb9-11"><a href="#cb9-11" tabindex="-1"></a>  <span class="fu">as.numeric</span>(<span class="fu">sub</span>(<span class="st">&quot;window_&quot;</span>, <span class="st">&quot;&quot;</span>, fragmentoverlap<span class="sc">$</span>window))</span>
<span id="cb9-12"><a href="#cb9-12" tabindex="-1"></a></span>
<span id="cb9-13"><a href="#cb9-13" tabindex="-1"></a>x <span class="ot">=</span> <span class="fu">match</span>(fragmentoverlap<span class="sc">$</span>barcode,</span>
<span id="cb9-14"><a href="#cb9-14" tabindex="-1"></a>        result<span class="sc">$</span>cellwindowCN<span class="sc">$</span>barcode)</span>
<span id="cb9-15"><a href="#cb9-15" tabindex="-1"></a>fragmentoverlap<span class="sc">$</span>CN <span class="ot">=</span> result<span class="sc">$</span>cellwindowCN<span class="sc">$</span>CN[x]</span>
<span id="cb9-16"><a href="#cb9-16" tabindex="-1"></a>fragmentoverlap<span class="sc">$</span>ploidy.moment.cell <span class="ot">=</span> result<span class="sc">$</span>cellwindowCN<span class="sc">$</span>ploidy.moment.cell[x]</span>
<span id="cb9-17"><a href="#cb9-17" tabindex="-1"></a>fragmentoverlap <span class="ot">=</span> fragmentoverlap[<span class="sc">!</span><span class="fu">is.na</span>(fragmentoverlap<span class="sc">$</span>CN), ]</span>
<span id="cb9-18"><a href="#cb9-18" tabindex="-1"></a></span>
<span id="cb9-19"><a href="#cb9-19" tabindex="-1"></a><span class="co"># For better hierarchical clustering</span></span>
<span id="cb9-20"><a href="#cb9-20" tabindex="-1"></a>fragmentoverlap<span class="sc">$</span>pwindownormalizedcleanedceiled <span class="ot">=</span></span>
<span id="cb9-21"><a href="#cb9-21" tabindex="-1"></a>  <span class="fu">pmin</span>(fragmentoverlap<span class="sc">$</span>CN, <span class="fu">min</span>(levels) <span class="sc">*</span> <span class="dv">2</span>)</span></code></pre></div>
<p>Make dataframe for plotting.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a>dataplot <span class="ot">=</span></span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a>  fragmentoverlap <span class="sc">%&gt;%</span></span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="st">&quot;w&quot;</span>, <span class="st">&quot;cell&quot;</span>, <span class="st">&quot;pwindownormalizedcleanedceiled&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">pivot_wider</span>(<span class="at">names_from =</span> <span class="st">&quot;w&quot;</span>, <span class="at">values_from =</span> <span class="st">&quot;pwindownormalizedcleanedceiled&quot;</span>)</span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a>dataplot <span class="ot">=</span> <span class="fu">as.data.frame</span>(dataplot)</span>
<span id="cb10-6"><a href="#cb10-6" tabindex="-1"></a><span class="fu">rownames</span>(dataplot) <span class="ot">=</span> dataplot<span class="sc">$</span>cell</span>
<span id="cb10-7"><a href="#cb10-7" tabindex="-1"></a>dataplot <span class="ot">=</span> dataplot[, <span class="fu">colnames</span>(dataplot) <span class="sc">!=</span> <span class="st">&quot;cell&quot;</span>]</span>
<span id="cb10-8"><a href="#cb10-8" tabindex="-1"></a>dataplot <span class="ot">=</span> <span class="fu">as.matrix</span>(dataplot)</span>
<span id="cb10-9"><a href="#cb10-9" tabindex="-1"></a>n <span class="ot">=</span> <span class="fu">max</span>(<span class="fu">as.numeric</span>(<span class="fu">colnames</span>(dataplot)))</span>
<span id="cb10-10"><a href="#cb10-10" tabindex="-1"></a>dataplot <span class="ot">=</span> dataplot[, <span class="fu">match</span>(<span class="fu">as.character</span>(<span class="dv">1</span><span class="sc">:</span>n), <span class="fu">colnames</span>(dataplot))]</span>
<span id="cb10-11"><a href="#cb10-11" tabindex="-1"></a><span class="fu">colnames</span>(dataplot) <span class="ot">=</span> <span class="fu">as.character</span>(<span class="dv">1</span><span class="sc">:</span>n)</span></code></pre></div>
<p>Plot.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a>breaks <span class="ot">=</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="fu">min</span>(levels) <span class="sc">-</span> <span class="dv">1</span>, <span class="fu">min</span>(levels) <span class="sc">+</span> <span class="dv">1</span>, <span class="fu">min</span>(levels) <span class="sc">*</span> <span class="dv">2</span>)</span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a></span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a>x <span class="ot">=</span> windowcovariates</span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a>x<span class="sc">$</span>chr[<span class="fu">duplicated</span>(windowcovariates<span class="sc">$</span>chr)] <span class="ot">=</span> <span class="cn">NA</span></span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a>x <span class="ot">=</span> x<span class="sc">$</span>chr[<span class="fu">match</span>(<span class="fu">colnames</span>(dataplot), x<span class="sc">$</span>w)]</span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a></span>
<span id="cb11-7"><a href="#cb11-7" tabindex="-1"></a>RowSideColors <span class="ot">=</span></span>
<span id="cb11-8"><a href="#cb11-8" tabindex="-1"></a>  <span class="fu">unlist</span>(</span>
<span id="cb11-9"><a href="#cb11-9" tabindex="-1"></a>    <span class="fu">lapply</span>(</span>
<span id="cb11-10"><a href="#cb11-10" tabindex="-1"></a>      fragmentoverlap<span class="sc">$</span>ploidy.moment.cell[</span>
<span id="cb11-11"><a href="#cb11-11" tabindex="-1"></a>        <span class="fu">match</span>(<span class="fu">rownames</span>(dataplot), fragmentoverlap<span class="sc">$</span>cell)],</span>
<span id="cb11-12"><a href="#cb11-12" tabindex="-1"></a>      <span class="cf">function</span> (x) { <span class="fu">which</span>(<span class="fu">sort</span>(levels) <span class="sc">==</span> x)}))</span>
<span id="cb11-13"><a href="#cb11-13" tabindex="-1"></a>RowSideColors <span class="ot">=</span> <span class="fu">topo.colors</span>(<span class="fu">length</span>(levels))[RowSideColors]</span>
<span id="cb11-14"><a href="#cb11-14" tabindex="-1"></a></span>
<span id="cb11-15"><a href="#cb11-15" tabindex="-1"></a>gplots<span class="sc">::</span><span class="fu">heatmap.2</span>(</span>
<span id="cb11-16"><a href="#cb11-16" tabindex="-1"></a>  dataplot,</span>
<span id="cb11-17"><a href="#cb11-17" tabindex="-1"></a>  <span class="at">Colv =</span> <span class="cn">FALSE</span>,</span>
<span id="cb11-18"><a href="#cb11-18" tabindex="-1"></a>  <span class="at">dendrogram =</span> <span class="st">&quot;none&quot;</span>,</span>
<span id="cb11-19"><a href="#cb11-19" tabindex="-1"></a>  <span class="at">breaks =</span> breaks,</span>
<span id="cb11-20"><a href="#cb11-20" tabindex="-1"></a>  <span class="at">col =</span> <span class="fu">c</span>(<span class="st">&quot;blue&quot;</span>, <span class="st">&quot;gray80&quot;</span>, <span class="st">&quot;red&quot;</span>),</span>
<span id="cb11-21"><a href="#cb11-21" tabindex="-1"></a>  <span class="at">trace =</span> <span class="st">&quot;none&quot;</span>, <span class="at">labRow =</span> <span class="cn">FALSE</span>, <span class="at">na.color =</span> <span class="st">&quot;white&quot;</span>,</span>
<span id="cb11-22"><a href="#cb11-22" tabindex="-1"></a>  <span class="at">labCol =</span> x,</span>
<span id="cb11-23"><a href="#cb11-23" tabindex="-1"></a>  <span class="at">RowSideColors=</span> RowSideColors)</span></code></pre></div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
