<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>Introduction to scPloidy</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Introduction to scPloidy</h1>



<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p><code>scPloidy</code> is a package to compute ploidy of single cells
(or nuclei) based on single-cell (or single-nucleus) ATAC-seq data. In
ATAC-seq, open chromatin regions are excised and sequenced. For any site
on the genome, ATAC-seq could read 0, 1 or 2 DNA fragments, if the cell
was diploid. If the cell was tetraploid, ATAC-seq could read 0, 1, 2, 3
or 4 fragments from the same site. This is the basic idea used in
<code>scPloidy</code>. We model the depth of DNA sequencing at one site
by binomial distribution.</p>
</div>
<div id="usage" class="section level2">
<h2>Usage</h2>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(scPloidy)</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">library</span>(GenomicRanges)</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="co">#&gt; Loading required package: stats4</span></span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="co">#&gt; Loading required package: BiocGenerics</span></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="co">#&gt; Attaching package: &#39;BiocGenerics&#39;</span></span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a><span class="co">#&gt; The following objects are masked from &#39;package:dplyr&#39;:</span></span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a><span class="co">#&gt;     combine, intersect, setdiff, union</span></span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a><span class="co">#&gt; The following objects are masked from &#39;package:stats&#39;:</span></span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb1-12"><a href="#cb1-12" tabindex="-1"></a><span class="co">#&gt;     IQR, mad, sd, var, xtabs</span></span>
<span id="cb1-13"><a href="#cb1-13" tabindex="-1"></a><span class="co">#&gt; The following objects are masked from &#39;package:base&#39;:</span></span>
<span id="cb1-14"><a href="#cb1-14" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb1-15"><a href="#cb1-15" tabindex="-1"></a><span class="co">#&gt;     Filter, Find, Map, Position, Reduce, anyDuplicated, aperm, append,</span></span>
<span id="cb1-16"><a href="#cb1-16" tabindex="-1"></a><span class="co">#&gt;     as.data.frame, basename, cbind, colnames, dirname, do.call,</span></span>
<span id="cb1-17"><a href="#cb1-17" tabindex="-1"></a><span class="co">#&gt;     duplicated, eval, evalq, get, grep, grepl, intersect, is.unsorted,</span></span>
<span id="cb1-18"><a href="#cb1-18" tabindex="-1"></a><span class="co">#&gt;     lapply, mapply, match, mget, order, paste, pmax, pmax.int, pmin,</span></span>
<span id="cb1-19"><a href="#cb1-19" tabindex="-1"></a><span class="co">#&gt;     pmin.int, rank, rbind, rownames, sapply, setdiff, table, tapply,</span></span>
<span id="cb1-20"><a href="#cb1-20" tabindex="-1"></a><span class="co">#&gt;     union, unique, unsplit, which.max, which.min</span></span>
<span id="cb1-21"><a href="#cb1-21" tabindex="-1"></a><span class="co">#&gt; Loading required package: S4Vectors</span></span>
<span id="cb1-22"><a href="#cb1-22" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb1-23"><a href="#cb1-23" tabindex="-1"></a><span class="co">#&gt; Attaching package: &#39;S4Vectors&#39;</span></span>
<span id="cb1-24"><a href="#cb1-24" tabindex="-1"></a><span class="co">#&gt; The following object is masked from &#39;package:gplots&#39;:</span></span>
<span id="cb1-25"><a href="#cb1-25" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb1-26"><a href="#cb1-26" tabindex="-1"></a><span class="co">#&gt;     space</span></span>
<span id="cb1-27"><a href="#cb1-27" tabindex="-1"></a><span class="co">#&gt; The following object is masked from &#39;package:tidyr&#39;:</span></span>
<span id="cb1-28"><a href="#cb1-28" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb1-29"><a href="#cb1-29" tabindex="-1"></a><span class="co">#&gt;     expand</span></span>
<span id="cb1-30"><a href="#cb1-30" tabindex="-1"></a><span class="co">#&gt; The following objects are masked from &#39;package:dplyr&#39;:</span></span>
<span id="cb1-31"><a href="#cb1-31" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb1-32"><a href="#cb1-32" tabindex="-1"></a><span class="co">#&gt;     first, rename</span></span>
<span id="cb1-33"><a href="#cb1-33" tabindex="-1"></a><span class="co">#&gt; The following object is masked from &#39;package:utils&#39;:</span></span>
<span id="cb1-34"><a href="#cb1-34" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb1-35"><a href="#cb1-35" tabindex="-1"></a><span class="co">#&gt;     findMatches</span></span>
<span id="cb1-36"><a href="#cb1-36" tabindex="-1"></a><span class="co">#&gt; The following objects are masked from &#39;package:base&#39;:</span></span>
<span id="cb1-37"><a href="#cb1-37" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb1-38"><a href="#cb1-38" tabindex="-1"></a><span class="co">#&gt;     I, expand.grid, unname</span></span>
<span id="cb1-39"><a href="#cb1-39" tabindex="-1"></a><span class="co">#&gt; Loading required package: IRanges</span></span>
<span id="cb1-40"><a href="#cb1-40" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb1-41"><a href="#cb1-41" tabindex="-1"></a><span class="co">#&gt; Attaching package: &#39;IRanges&#39;</span></span>
<span id="cb1-42"><a href="#cb1-42" tabindex="-1"></a><span class="co">#&gt; The following objects are masked from &#39;package:dplyr&#39;:</span></span>
<span id="cb1-43"><a href="#cb1-43" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb1-44"><a href="#cb1-44" tabindex="-1"></a><span class="co">#&gt;     collapse, desc, slice</span></span>
<span id="cb1-45"><a href="#cb1-45" tabindex="-1"></a><span class="co">#&gt; Loading required package: GenomeInfoDb</span></span>
<span id="cb1-46"><a href="#cb1-46" tabindex="-1"></a><span class="fu">library</span>(IRanges)</span>
<span id="cb1-47"><a href="#cb1-47" tabindex="-1"></a><span class="fu">library</span>(readr)</span></code></pre></div>
<p>See description.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a>?fragmentoverlapcount</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>?ploidy</span></code></pre></div>
</div>
<div id="analyzing-sample-data" class="section level2">
<h2>Analyzing sample data</h2>
<p>In this section, we demonstrate the package by using a dataset
included in the package. See next section on how to analyze your own
data.</p>
<div id="compute-overlaps-of-dna-fragments" class="section level3">
<h3>Compute overlaps of DNA fragments</h3>
<p>We use small dataset for single-nucleus ATAC-seq of rat liver. To
limit the file size, it only includes 10 cells and fragments from
chromosomes 19 and 20. The fragments were mapped to the rat rn7
genome.</p>
<p>We first set the regions where overlaps are counted. This is usually
all of the autosomes.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a>targetregions <span class="ot">=</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>  GenomicRanges<span class="sc">::</span><span class="fu">GRanges</span>(</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>    <span class="fu">c</span>(<span class="st">&quot;chr19&quot;</span>, <span class="st">&quot;chr20&quot;</span>),</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>    IRanges<span class="sc">::</span><span class="fu">IRanges</span>(<span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">1</span>), <span class="at">width =</span> <span class="dv">500000000</span>))</span></code></pre></div>
<p>Simple repeats in the genome can generate false overlaps. We exclude
such regions. The regions were downloaded from USCS genome browser.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a>simpleRepeat <span class="ot">=</span> readr<span class="sc">::</span><span class="fu">read_tsv</span>(</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>  <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;simpleRepeat.chr19_20.txt.gz&quot;</span>, <span class="at">package =</span> <span class="st">&quot;scPloidy&quot;</span>),</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a>  <span class="at">col_names =</span> <span class="fu">c</span>(<span class="st">&quot;chrom&quot;</span>, <span class="st">&quot;chromStart&quot;</span>, <span class="st">&quot;chromEnd&quot;</span>))</span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a><span class="co">#&gt; Rows: 107037 Columns: 3</span></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a><span class="co">#&gt; ── Column specification ────────────────────────────────────────────────────────</span></span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a><span class="co">#&gt; Delimiter: &quot;\t&quot;</span></span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a><span class="co">#&gt; chr (1): chrom</span></span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a><span class="co">#&gt; dbl (2): chromStart, chromEnd</span></span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a><span class="co">#&gt; ℹ Use `spec()` to retrieve the full column specification for this data.</span></span>
<span id="cb4-11"><a href="#cb4-11" tabindex="-1"></a><span class="co">#&gt; ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</span></span>
<span id="cb4-12"><a href="#cb4-12" tabindex="-1"></a>simpleRepeat[, <span class="dv">2</span>] <span class="ot">=</span> simpleRepeat[, <span class="dv">2</span>] <span class="sc">+</span> <span class="dv">1</span> <span class="co"># convert from 0-based position to 1-based</span></span>
<span id="cb4-13"><a href="#cb4-13" tabindex="-1"></a>simpleRepeat <span class="ot">=</span> GenomicRanges<span class="sc">::</span><span class="fu">makeGRangesFromDataFrame</span>(</span>
<span id="cb4-14"><a href="#cb4-14" tabindex="-1"></a>  <span class="fu">as.data.frame</span>(simpleRepeat),</span>
<span id="cb4-15"><a href="#cb4-15" tabindex="-1"></a>  <span class="at">seqnames.field =</span> <span class="st">&quot;chrom&quot;</span>,</span>
<span id="cb4-16"><a href="#cb4-16" tabindex="-1"></a>  <span class="at">start.field    =</span> <span class="st">&quot;chromStart&quot;</span>,</span>
<span id="cb4-17"><a href="#cb4-17" tabindex="-1"></a>  <span class="at">end.field      =</span> <span class="st">&quot;chromEnd&quot;</span>)</span></code></pre></div>
<p>Now compute the overlaps. The input file
<code>SHR_m154211.10cells.chr19_20.fragments.txt.gz</code> records the
fragments sequenced in ATAC-seq.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a>fragmentoverlap <span class="ot">=</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>  <span class="fu">fragmentoverlapcount</span>(</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a>    <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;SHR_m154211.10cells.chr19_20.fragments.txt.gz&quot;</span>, <span class="at">package =</span> <span class="st">&quot;scPloidy&quot;</span>),</span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a>    targetregions,</span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a>    <span class="at">excluderegions =</span> simpleRepeat,</span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a>    <span class="at">Tn5offset =</span> <span class="fu">c</span>(<span class="dv">4</span>, <span class="sc">-</span><span class="dv">5</span>))</span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a><span class="co">#&gt;   |                                                                              |                                                                      |   0%  |                                                                              |===================================                                   |  50%  |                                                                              |======================================================================| 100%</span></span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a>fragmentoverlap</span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a><span class="co">#&gt; # A tibble: 10 × 8</span></span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a><span class="co">#&gt;    barcode     nfrags depth1 depth2 depth3 depth4 depth5 depth6</span></span>
<span id="cb5-11"><a href="#cb5-11" tabindex="-1"></a><span class="co">#&gt;    &lt;chr&gt;        &lt;int&gt;  &lt;int&gt;  &lt;int&gt;  &lt;int&gt;  &lt;int&gt;  &lt;int&gt;  &lt;int&gt;</span></span>
<span id="cb5-12"><a href="#cb5-12" tabindex="-1"></a><span class="co">#&gt;  1 BC00019_N01   6501   6288    201      9      2      1      0</span></span>
<span id="cb5-13"><a href="#cb5-13" tabindex="-1"></a><span class="co">#&gt;  2 BC00022_N01   8006   7689    310      6      1      0      0</span></span>
<span id="cb5-14"><a href="#cb5-14" tabindex="-1"></a><span class="co">#&gt;  3 BC00025_N01   5904   5767    133      3      1      0      0</span></span>
<span id="cb5-15"><a href="#cb5-15" tabindex="-1"></a><span class="co">#&gt;  4 BC00026_N01   5257   5102    148      6      1      0      0</span></span>
<span id="cb5-16"><a href="#cb5-16" tabindex="-1"></a><span class="co">#&gt;  5 BC00035_N01   6360   6198    158      4      0      0      0</span></span>
<span id="cb5-17"><a href="#cb5-17" tabindex="-1"></a><span class="co">#&gt;  6 BC00055_N01   6271   6084    181      6      0      0      0</span></span>
<span id="cb5-18"><a href="#cb5-18" tabindex="-1"></a><span class="co">#&gt;  7 BC00057_N01   5570   5413    152      4      1      0      0</span></span>
<span id="cb5-19"><a href="#cb5-19" tabindex="-1"></a><span class="co">#&gt;  8 BC00077_N01   5997   5795    191     10      1      0      0</span></span>
<span id="cb5-20"><a href="#cb5-20" tabindex="-1"></a><span class="co">#&gt;  9 BC00086_N01   5416   5261    151      4      0      0      0</span></span>
<span id="cb5-21"><a href="#cb5-21" tabindex="-1"></a><span class="co">#&gt; 10 BC00087_N01   3470   3409     60      1      0      0      0</span></span>
<span id="cb5-22"><a href="#cb5-22" tabindex="-1"></a><span class="fu">rm</span>(fragmentoverlap)</span></code></pre></div>
</div>
<div id="infer-ploidy" class="section level3">
<h3>Infer ploidy</h3>
<p>Instead of the small dataset computed above, we here use a complete
dataset for single-nucleus ATAC-seq of a rat liver.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="fu">data</span>(SHR_m154211)</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>?SHR_m154211</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a>fragmentoverlap <span class="ot">=</span> SHR_m154211<span class="sc">$</span>fragmentoverlap</span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a>fragmentoverlap</span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a><span class="co">#&gt; # A tibble: 3,572 × 8</span></span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a><span class="co">#&gt;    barcode     nfrags depth1 depth2 depth3 depth4 depth5 depth6</span></span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a><span class="co">#&gt;    &lt;chr&gt;        &lt;int&gt;  &lt;int&gt;  &lt;int&gt;  &lt;int&gt;  &lt;int&gt;  &lt;int&gt;  &lt;int&gt;</span></span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a><span class="co">#&gt;  1 BC00025_N01 129201 126908   2237     53      3      0      0</span></span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a><span class="co">#&gt;  2 BC00026_N01 130564 127401   3065     91      6      1      0</span></span>
<span id="cb6-10"><a href="#cb6-10" tabindex="-1"></a><span class="co">#&gt;  3 BC00035_N01 123100 120093   2940     66      1      0      0</span></span>
<span id="cb6-11"><a href="#cb6-11" tabindex="-1"></a><span class="co">#&gt;  4 BC00057_N01 110602 108061   2472     67      2      0      0</span></span>
<span id="cb6-12"><a href="#cb6-12" tabindex="-1"></a><span class="co">#&gt;  5 BC00086_N01 110108 107374   2667     66      1      0      0</span></span>
<span id="cb6-13"><a href="#cb6-13" tabindex="-1"></a><span class="co">#&gt;  6 BC00087_N01 108661 106717   1908     33      2      1      0</span></span>
<span id="cb6-14"><a href="#cb6-14" tabindex="-1"></a><span class="co">#&gt;  7 BC00099_N01 107028 104685   2264     76      3      0      0</span></span>
<span id="cb6-15"><a href="#cb6-15" tabindex="-1"></a><span class="co">#&gt;  8 BC00106_N01 100659  97576   2985     92      6      0      0</span></span>
<span id="cb6-16"><a href="#cb6-16" tabindex="-1"></a><span class="co">#&gt;  9 BC00115_N01 105057 102889   2104     64      0      0      0</span></span>
<span id="cb6-17"><a href="#cb6-17" tabindex="-1"></a><span class="co">#&gt; 10 BC00116_N01 105085 103519   1535     31      0      0      0</span></span>
<span id="cb6-18"><a href="#cb6-18" tabindex="-1"></a><span class="co">#&gt; # ℹ 3,562 more rows</span></span></code></pre></div>
<p>Compute ploidy, assuming a mixture of 2x (diploid), 4x (tetraploid)
and 8x cells. We recommend using <code>ploidy.moment</code> which is
based on moment method.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a>p <span class="ot">=</span> <span class="fu">ploidy</span>(fragmentoverlap,</span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>           <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">4</span>, <span class="dv">8</span>))</span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a><span class="co">#&gt; number of iterations= 178</span></span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a><span class="fu">head</span>(p)</span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a><span class="co">#&gt;       barcode ploidy.moment ploidy.momentfractional ploidy.kmeans ploidy.em</span></span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a><span class="co">#&gt; 1 BC00025_N01             8                5.706704             4         4</span></span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a><span class="co">#&gt; 2 BC00026_N01             8                7.485560             8         8</span></span>
<span id="cb7-8"><a href="#cb7-8" tabindex="-1"></a><span class="co">#&gt; 3 BC00035_N01             8                7.391306             4         4</span></span>
<span id="cb7-9"><a href="#cb7-9" tabindex="-1"></a><span class="co">#&gt; 4 BC00057_N01             8                7.083094             8         8</span></span>
<span id="cb7-10"><a href="#cb7-10" tabindex="-1"></a><span class="co">#&gt; 5 BC00086_N01             8                7.522005             8         4</span></span>
<span id="cb7-11"><a href="#cb7-11" tabindex="-1"></a><span class="co">#&gt; 6 BC00087_N01             8                5.703602             4         4</span></span></code></pre></div>
<p>The cell type of the cells are stored in dataframe
<code>cells</code>. There are many hepatocytes of 4x and 8x, but other
cell types are mostly 2x.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a>cells <span class="ot">=</span> SHR_m154211<span class="sc">$</span>cells</span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a><span class="fu">table</span>(cells<span class="sc">$</span>celltype,</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a>      p<span class="sc">$</span>ploidy.moment[<span class="fu">match</span>(cells<span class="sc">$</span>barcode, p<span class="sc">$</span>barcode)])</span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a><span class="co">#&gt;              </span></span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a><span class="co">#&gt;                  2    4    8</span></span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a><span class="co">#&gt;   B             31   24    2</span></span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a><span class="co">#&gt;   T_NK          65   55    1</span></span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a><span class="co">#&gt;   endothelial  384  196    2</span></span>
<span id="cb8-9"><a href="#cb8-9" tabindex="-1"></a><span class="co">#&gt;   hepatocyte   203 1343  780</span></span>
<span id="cb8-10"><a href="#cb8-10" tabindex="-1"></a><span class="co">#&gt;   macrophage   140  139    1</span></span>
<span id="cb8-11"><a href="#cb8-11" tabindex="-1"></a><span class="co">#&gt;   stellate     107   93    6</span></span></code></pre></div>
</div>
</div>
<div id="analyzing-your-own-data" class="section level2">
<h2>Analyzing your own data</h2>
<div id="using-fragments.tsv.gz-generated-by-10x-cell-ranger" class="section level3">
<h3>Using fragments.tsv.gz generated by 10x Cell Ranger</h3>
<p>In the Cell Ranger output directory, you should have files
<code>fragments.tsv.gz</code> and <code>fragments.tsv.gz.tbi</code>. The
file <code>fragments.tsv.gz</code> can be processed with the
<code>fragmentoverlapcount</code> function, specifying the option
<code>Tn5offset = c(1, 0)</code></p>
</div>
<div id="using-bam-file" class="section level3">
<h3>Using BAM file</h3>
<p>Although this requires several steps, you can start from a BAM file
and generate fragments file for <code>scPloidy</code>. You need to
install <code>samtools</code>, <code>bgzip</code> and
<code>tabix</code>, and run the following in your shell.</p>
<p>First generate a BAM file in which the cell barcode is prepended to
read name, separated by ‘:’. For example, suppose in your input file
<code>bap.bam</code> your barcode <code>BCxxxxxx</code> was stored in
the field with <code>DB</code> tag as
<code>DB:Z:atac_possorted_bam_BCxxxxxx</code>. We generate a BAM file
<code>snap.bam</code>.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="co"># extract the header file</span></span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a><span class="ex">samtools</span> view bap.bam <span class="at">-H</span> <span class="op">&gt;</span> header.sam</span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a></span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a><span class="co"># create a bam file with the barcode embedded into the read name</span></span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a><span class="fu">cat</span> <span class="op">&lt;(</span> <span class="fu">cat</span> header.sam <span class="op">)</span> <span class="dt">\</span></span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a><span class="op">&lt;(</span> <span class="ex">samtools</span> view bap.bam <span class="kw">|</span> <span class="fu">awk</span> <span class="st">&#39;{for (i=12; i&lt;=NF; ++i) { if ($i ~ &quot;^DB:Z:atac_possorted_bam_BC&quot;){ td[substr($i,1,2)] = substr($i,25,length($i)-24); } }; printf &quot;%s:%s\n&quot;, td[&quot;DB&quot;], $0 }&#39;</span> <span class="op">)</span> <span class="dt">\</span></span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a><span class="kw">|</span> <span class="ex">samtools</span> view <span class="at">-bS</span> <span class="at">-</span> <span class="op">&gt;</span> snap.bam</span></code></pre></div>
<p>We next sort the reads by barcode, and obtain the file
<code>snap.nsrt.bam</code>.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="ex">samtools</span> sort <span class="at">-n</span> <span class="at">-@</span> 20 <span class="at">-m</span> 10G snap.bam <span class="at">-o</span> snap.nsrt.bam</span></code></pre></div>
<p>Finally, we extract fragments from the name-sorted BAM file, and
obtain the file <code>fragments.txt.gz</code>.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="ex">samtools</span> view <span class="at">-f</span> 0x2 <span class="at">-F</span> 0x900 <span class="at">-q</span> 30 snap.nsrt.bam  <span class="dt">\</span></span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a>  <span class="kw">|</span> <span class="ex">samtofragmentbed.pl</span>  <span class="dt">\</span></span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a>  <span class="kw">|</span> <span class="fu">perl</span> <span class="at">-ne</span> <span class="st">&#39;chomp; @a=split; $a[3]=~s/:.*//; print join(&quot;\t&quot;,@a), &quot;\n&quot;&#39;</span>  <span class="dt">\</span></span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a>  <span class="kw">|</span> <span class="va">LC_ALL</span><span class="op">=</span>C <span class="fu">sort</span> <span class="at">-T</span> ./ <span class="at">-S</span> 50% <span class="at">--parallel</span><span class="op">=</span>12 <span class="at">-k1,1</span> <span class="at">-k2,3n</span> <span class="at">-k4,4</span> <span class="at">-t</span><span class="st">$&#39;</span><span class="dt">\t</span><span class="st">&#39;</span> <span class="kw">|</span> <span class="fu">uniq</span>  <span class="dt">\</span></span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a>  <span class="kw">|</span> <span class="ex">bgzip</span> <span class="op">&gt;</span> fragments.txt.gz</span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a><span class="ex">tabix</span> <span class="at">-b</span> 2 <span class="at">-e</span> 3 fragments.txt.gz</span></code></pre></div>
<p>The Perl script <code>samtofragmentbed.pl</code> is included in this
package as this file:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="fu">system.file</span>(<span class="st">&quot;perl&quot;</span>, <span class="st">&quot;samtofragmentbed.pl&quot;</span>, <span class="at">package =</span> <span class="st">&quot;scPloidy&quot;</span>)</span></code></pre></div>
<p>The file <code>fragments.txt.gz</code> can be processed with the
<code>fragmentoverlapcount</code> function, specifying the option
<code>Tn5offset = c(4, -5)</code></p>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
